### ISSUE TITLE
new branch type: "prototype" 

### ISSUE THREAD

**Issue by kevgo** [git-town/git-town#3646]
A prototype branch is a local-only feature branch that incorporates updates from its parent branch but is not pushed to the remote repository. This branch type is particularly useful in several scenarios:

- The branch contains sensitive information, such as secrets, or potentially problematic code or data that could trigger alerts.
- The developer prefers to keep their work private from the rest of the team during the initial stages of development.
- Prototype branches can help conserve CI minutes during the early phases of feature development, although ticket #999 addresses this issue more specifically.

### Example workflow

```bash
# let's create a new prototype-branch
> git prototype my-branch
# now we are on the "my-branch" branch
# (hardcore hacking here)
> git sync
# doesn't create a tracking branch, just pulls in the parent
> git propose
# creates a tracking branch, opens the browser for creating a PR
# adding a tracking branch makes this branch a normal feature branch
> git sync
# now it pushes local commits to the remote
```

### action items

- [x] `git prototype` - make the current branch a prototype branch
- [x] `git prototype existing-branch-name` - make the given branch a prototype
- [x] `git hack --prototype new-branch-name` - create a new top-level prototype branch
- [x] `git append --prototype new-branch-name` - create a new prototype branch as a child of the current branch
- [x] `git prepend --prototype new-branch-name` - create a new prototype branch between the current branch and its parent
- [x] `git propose` on a prototype branch converts the branch to a regular feature branch
- [x] `git hack existing-prototype-branch-name` - convert the given prototype branch into a regular feature branch
- [x] Git setting `git-town.create-prototype-branches` - if enabled, all new branches are created as prototype branches
- [ ] git-town.sync-prototype-strategy setting

**Comment by stephenwade** [comment]
This would fit my desired workflow perfectly!

Would `git hack` and `git prototype` convert a branch between types, just like `git observe`?

**Comment by kevgo** [comment]
Alternative ways to create the various forms of prototype branch that Git Town allows:

```bash
git hack --prototype new-branch
git append --prototype new-branch
git prepend --prototype new-branch
```

Then the actual `prototype` commands would do:

```bash
# makes the current branch a prototype branch
git prototype
# makes the given branch a prototype branch
git prototype existing-branch
# creates a new prototype branch off the main branch
git prototype new-branch
```

The last two commands seem optional.

**Comment by stephenwade** [comment]
I would love a setting to make new branches prototype branches by default.

**Comment by kevgo** [comment]
Yes, such a setting would make sense.

**Comment by kevgo** [comment]
- `git prototype` - make the current branch a prototype branch
- `git prototype branch-name` - make the given branch a prototype
- `git hack --prototype branch-name` - create a new top-level prototype branch
- `git append --prototype branch-name` - create a new prototype branch as a child of the current branch
- `git prepend --prototype branch-name` - create a new prototype branch between the current branch and its parent

**Comment by kevgo** [comment]
I think `git prototype new-branch` should not do anything. 

1. It is completely redundant with `git hack --prototype new-branch` or `git append --prototype new-branch`. Redundant ways to do the same thing are an API smell.
2. It's unclear whether it _will_ create the new prototype branch as a top-level branch or as a child of the current branch. This would be bad API design, resulting in an API that is hard to understand and easy to use wrong.
3. It's equally unclear which of the two options it _should_ do. Making this a good API isn't possible.

Given (1), (2), (3) are true together, it's best to not do anything and force the user to use `git hack`, `git append`, or `git prepend` with the `--prototype` switch.

**Comment by kevgo** [comment]
Possible names for the boolean config setting to automatically create prototype branches:

1. `git-town.always-create-prototype-branches = true`
2. `git-town.always-prototype = true`
3. `git-town.auto-prototype = true`
4. `git-town.prototype-branches = always`
5. `git-town.prototypes = always`
6. `git-town.prototype = always`

**Comment by kevgo** [comment]
I think `create-prototype-branches` is the most accurate. The other settings sound too vague or too broad.

**Comment by stephenwade** [comment]
**suggestion:** Add the setting `git-town.sync-perennial-strategy = <merge|rebase>`

I want to rebase my prototype branches on `main` before proposing them, but I want to use merges to sync all other branches so I don't have to force push to the remote.

**Comment by kevgo** [comment]
`git town propose` now also converts a prototype branch to a regular feature branch. `git sync` syncs prototype branches only locally.

**Comment by kevgo** [comment]
This is now available in Git Town [v15.0](https://github.com/git-town/git-town/releases/tag/v15.0.0).
